python-lib/ontology_tagging/formatter_by_document.py

```diff
@@ -0,0 +1,442 @@
+            self.output_df.set_index(column, inplace=True)
+            self.output_df.reset_index(inplace=True)
+        return self._set_columns_order(input_df, self.output_df, text_column)
+
+
+class FormatterByDocumentJson(FormatterByDocument):
+    """Inherits from FormatterBase class.
+    Class to write a dataframe which contains one row per document, with tag datas in nested json columns
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(FormatterByDocumentJson, self).__init__(*args, **kwargs)
+
+    def write_df(
+        self,
+        input_df: pd.DataFrame,
+        text_column: AnyStr,
+        language_column: AnyStr = None,
+    ) -> pd.DataFrame:
+        """Write the output dataframe for format 'one row per document (nested JSON)' when there is no category in the ontology
+
+        Returns:
+            pd.DataFrame : a dataframe with the following columns:
+            - all columns from the input dataframe input_df
+            - a column with a dictionary of tags (keys). Each tag has for value a dictionary with:
+                -key 'count' : number of occurences of the keywords in the document
+                -key 'sentences' : list of the sentences that matched keywords in the document
+                -key 'keywords' : list of the keywords in the document
+
+        """
+        start = perf_counter()
+        self._generate_columns_names(input_df)
+        input_df.progress_apply(self._write_row, args=[language_column], axis=1)
+        logging.info(
+            f"Tagging {len(input_df)} documents : Done in {perf_counter() - start:.2f} seconds."
+        )
+        return self._set_columns_order(input_df, self.output_df, text_column)
+
+    def _write_row(self, row: pd.Series, language_column: AnyStr = None) -> None:
+        """
+        Called by write_df on each row
+        Update the column 'tag_json_full' with the found tags
+
+        Args:
+            row: pandas.Series from text_df
+            language_column: if not None, matcher will be applied with the given language of the row; uses self.language otherwise
+
+        """
+        language = self._get_document_language(row, language_column)
+        document_to_match = self._get_document_to_match(row, language)
+        line_full, tag_column_for_json = defaultdict(defaultdict), {}
+        for idx, sentence in enumerate(document_to_match):
+            matches = self._matcher_dict[language](sentence, as_spans=True)
+            for keyword in matches:
+                line_full = self._get_tags_in_row(
+                    match=keyword,
+                    line_full=line_full,
+                    original_sentence=row[self.text_column_tokenized][idx],
+                    language=language,
+                )
+
+        tag_column_for_json["tag_json_full"] = self._fill_tags(
+            condition=line_full,
+            value={
+                column_name: dict(value) for column_name, value in line_full.items()
+            },
```

Comment:
Why not simply ?
```suggestion
            value=line_full,
```

Reply:
Because it displays rows "defaultdict(...)" in the output dataset otherwise